doc: Protocols: added Wi-Fi documentation (#9894)
nrfconnect/sdk-nrf: doc/nrf/ug_wifi.rst

.. _ug_wifi:

Wi-Fi
#####

.. contents::
   :local:
   :depth: 2

> Wi-Fi is the branding used by the Wi-Fi Alliance to describe the family of wireless local area network standards specified by the IEEE802.11 Working Group.

Reviewer: Hyperlinks to Wi-Fi Alliance and IEEE802.11 WG.
https://www.wi-fi.org/
https://www.ieee802.org/11/

The two industry bodies work closely together, with the Wi-Fi Alliance focusing on the requirements and certification of the technology, and the IEEE focusing on the underlying technical specifications.

> The Wi-Fi Alliance is responsible for the Wi-Fi CERTIFIED accreditation program, which aims to achieve high level of interoperability, reliability, and security compliance.

Reviewer: ```suggestion
The Wi-Fi Alliance is responsible for the Wi-Fi CERTIFIED accreditation program, which aims to achieve high level of interoperability, reliability, and security compliance. Further information related to certification on the nRF70 Series can be found here [link to infocenter WFA certification doc].
```

Author: Infocenter docs are not available yet.

Reviewer: The infocenter structure for guidelines is still being worked out. This will not be known until Wed/Thurs this week.

Wi-Fi is an evolving standard, with new IEEE802.11 standards being incorporated every 4-6 years, along with the corresponding WFA certification program.
The branding “Wi-Fi 6” is defined by the WFA, and aligns with the IEEE802.11ax specification, but includes all previous versions of the IEEE802.11 specifications dating back to the introduction of Wi-Fi in 1997.
The evolution from the base standard started with the introduction of IEEE802.11b and IEEE802.11a in 1999, IEEE802.11g in 2003, IEEE802.11n in 2008 (and subsequently branded Wi-Fi 4), IEEE802.11ac in 2014 (subsequently branded Wi-Fi 5), and finally IEEE802.11ax in 2019, branded Wi-Fi 6.
These standards operate in the unlicensed spectrum of the 2.4 GHz and 5 GHz bands.
However, 11b and 11g are only applicable in the 2.4 GHz band, while 11a and 11ac are only applicable in the 5 GHz band.
The IEEE802.11ax standard is also specified for operation in the 6 GHz band, with this variant being branded Wi-Fi 6E to signify extension to the 2.4 and 5 GHz bands used in Wi-Fi 6.
Wi-Fi 7, based on the IEEE802.11be specification, is due to launch by the end of 2023.

> 

Reviewer: ```suggestion
The features supported by nRF70 Series Wi-Fi devices can be found in [link to PSes in infocenter]
```

Author: Not available yet.

Reviewer: This will be the infocenter landing page for the Product Specs, https://infocenter.nordicsemi.com/topic/ps_nrf7002/keyfeatures_html5.html. This will be available by launch.

Author: Added a link.

Since the radio frequency (RF) communication channel is unreliable, a number of mechanisms are used in an attempt to achieve reliability not typically present in wired networks.

Firstly, multiple modulation schemes are used, with selection dependent on the prevailing RF channel conditions.
Secondly, forward error protection (FEC) is used when the data is encoded.
This adds redundancy which can be used at the receiver in an attempt to recover the original data content.

Multiple schemes (and rates) are available, depending on the version of the standard in use and the channel conditions.
Acknowledgements (ACKs) are used for unicast packet transmissions, with every successful packet reception resulting in an ACK transmission back to the originator confirming reception.

The RF medium also necessitates security to achieve privacy of the over-the-air packet transmissions.
Again, multiple schemes are available depending on the version of the standard in use, though over time the older schemes tend to be deprecated as vulnerabilities emerge.
The schemes are referred to as Wi-Fi Protected Access (WPA), with WPA3 being the latest version (mandated for all new WFA certifications).

> Protocol stack

Reviewer: ```suggestion
Wi-Fi protocol stack
```

**************

The IEEE802.11 specifications only define the PHY and MAC layers of the Wi-Fi protocol, with the remaining upper layers being common with other IP-based technologies, in particular Ethernet (IEEE802.3).

The following diagram details the full Wi-Fi protocol stack.

.. figure:: /images/wifi_protocol_stack.svg
   :alt: Wi-Fi protocol stack

   Wi-Fi protocol stack

> Networking stack

Reviewer: ```suggestion
Network stack
```

================

> At the top of the stack are application orientated network protocols, for example MQTT and LwM2M.

Reviewer: ```suggestion
At the top of the stack are application orientated network protocols, for example Matter, MQTT and LwM2M.
```

These are agnostic with respect to the underlying communication protocols.
The next level down is the TCP/IP stack, common to all IP-based communication protocols, such as Wi-Fi, Ethernet, and Thread.

> Wi-Fi driver

Reviewer: ```suggestion
Wi-Fi host stack
```

============

> Working down the stack, the first Wi-Fi specific component is the supplicant.
> It is primarily responsible for implementing the authentication phase of WPA (including WPA2 and WPA3).
> Authentication deals with security before association, and is typically divided into two categories of algorithm, personal and enterprise.
> 
> Personal authentication uses the pre-shared key (PSK) approach, which to the end user is usually presented as a network passphrase.
> 
> Enterprise security uses the 802.1X authentication protocol, and typically involves user certificates and a centralised authentication server (for example RADIUS).
> 
> Apart from WPA authentication, the supplicant performs other duties in Wi-Fi associated with joining a network such as network scanning, key exchange, and roaming support.

Reviewer: ```suggestion
Beneath the common network stack is the first Wi-Fi centric layer, the Wi-Fi host stack. There are two logical parts to the Wi-Fi host stack, namely the Wi-Fi driver and the supplicant.

Wi-Fi driver
------------

The Wi-Fi driver is responsible for presenting APIs for interfacing with the network stack and the supplicant, as well as handling communication with the Wi-Fi baseband, including managing any physical transport layers (e.g. SPI, USB, etc).

Supplicant
----------

The supplicant is primarily responsible for implementing the authentication phase of WPA (inc. WPA2/WPA3).
Authentication deals with security before association, and is typically divided into two categories of algorithm, Personal and Enterprise.
Personal authentication uses the pre-shared key (PSK) approach, which to the end user is usually presented as a network passphrase.
Enterprise security on the other hand uses the 802.1X authentication protocol, and typically involves user certificates and a centralised authentication server (e.g. RADIUS). Apart from WPA authentication, the supplicant performs other duties in Wi-Fi associated with joining a network such as network scanning, key exchange, roaming support, etc.
The supplicant used with nRF70 series devices is a Zephyr port of wpa_supplicant (https://w1.fi/wpa_supplicant/).
```

> Wi-Fi MAC layer

Reviewer: ```suggestion
MAC layer
```

WPA3 introduced an enhanced authentication scheme known as Simultaneous Authentication of Equals (SAE).
WPA3 is now mandated, and is a prerequisite for WFA certification.

Power save is the mechanism to enable network clients (STAs) to periodically enter a very low power state whereby they can shut down the majority of the transceiver functionality without risking any missed traffic.
This has traditionally been coordinated by the AP using a mechanism referred to as Delivery Traffic Indication Message (DTIM).
This message is a periodic indication to every STA on the network of whether there is pending receive traffic.
It is carried in a predetermined subset of the beacons so that a connected STA only needs to wake up to receive this message and based on this information, either return to sleep or initiate receiving of the pending traffic buffered by the AP.
With the advent of Wi-Fi 6, an alternative power save mechanism, known as Target Wake Time (TWT), has been introduced .
This mechanism also enables STAs to enter a low power sleep state with receive traffic being buffered at the AP, but with much more flexibility in terms of the sleep interval.

> Wi-Fi PHY layer

Reviewer: ```suggestion
Physical layer
```

Multiple schemes (and rates) are available, depending on the version of the standard in use and the channel conditions.
Acknowledgements (ACKs) are used for unicast packet transmissions, with every successful packet reception resulting in an ACK transmission back to the originator confirming reception.

The RF medium also necessitates security to achieve privacy of the over-the-air packet transmissions.
Again, multiple schemes are available depending on the version of the standard in use, though over time the older schemes tend to be deprecated as vulnerabilities emerge.
The schemes are referred to as Wi-Fi Protected Access (WPA), with WPA3 being the latest version (mandated for all new WFA certifications).

Wi-Fi Protocol stack
********************

The IEEE802.11 specifications only define the PHY and MAC layers of the Wi-Fi protocol, with the remaining upper layers being common with other IP-based technologies, in particular Ethernet (IEEE802.3).

The following diagram details the full Wi-Fi protocol stack.

.. figure:: /images/wifi_protocol_stack.svg
   :alt: Wi-Fi protocol stack

   Wi-Fi protocol stack

Network stack
=============

> At the top of the stack are application orientated network protocols, for example Matter, MQTT and LwM2M.

Reviewer 2: ```suggestion
At the top of the stack are application orientated network protocols, for example Matter, MQTT, and LwM2M.
```

There are two logical parts to the Wi-Fi host stack, namely the Wi-Fi driver and the supplicant.

Wi-Fi driver
------------
The Wi-Fi driver is responsible for presenting APIs for interfacing with the network stack and the supplicant, as well as handling communication with the Wi-Fi baseband, including managing any physical transport layers, for example SPI and USB.

Supplicant
----------
The supplicant is primarily responsible for implementing the authentication phase of WPA (WPA2/WPA3).
Authentication deals with security before association, and is typically divided into two categories of algorithm, Personal and Enterprise.
Personal authentication uses the pre-shared key (PSK) approach, which to the end user is usually presented as a network passphrase.
Enterprise security uses the 802.1X authentication protocol, and typically involves user certificates and a centralized authentication server, for example RADIUS.
Apart from WPA authentication, the supplicant performs other duties in Wi-Fi associated with joining a network such as network scanning, key exchange, and roaming support.
The supplicant used with nRF70 Series devices is a Zephyr port of `WPA Supplicant`_.

> MAC layer
> =========

Reviewer: I want to make the text heading levels follow the diagram more closely, in particular have the major headings aligned with the black column on the rhs. Then we can lose some of the text such as "continuing down the stack", since it will be easier to cross reference with the diagram.

```suggestion
Wi-Fi baseband
=========

The Wi-Fi baseband consists of the bottom two layers of the protocol stack, namely the Medium Access Control (MAC) layer and the Physical (PHY) layer. These two layers are fully specified in the IEEE802.11 standard.

MAC layer
---------
```

> Continuing down the stack, the next component is the Medium Access Control (MAC) layer.

Reviewer: ```suggestion
```

Packet level security deals with performing symmetric encryption at the MAC frame level.
The encryption algorithm in WPA is known as TKIP and uses the RC4 cipher, while WPA2 and WPA3 use encryption algorithms based on the AES cipher, with either 128 or 256-bit key lengths.
WPA3 introduced an enhanced authentication scheme known as Simultaneous Authentication of Equals (SAE).
WPA3 is now mandated, and is a prerequisite for WFA certification.

Power save is the mechanism to enable network clients (STAs) to periodically enter a very low power state whereby they can shut down the majority of the transceiver functionality without risking any missed traffic.
This has traditionally been coordinated by the AP using a mechanism referred to as Delivery Traffic Indication Message (DTIM).
This message is a periodic indication to every STA on the network of whether there is pending receive traffic.
It is carried in a predetermined subset of the beacons so that a connected STA only needs to wake up to receive this message and based on this information, either return to sleep or initiate receiving of the pending traffic buffered by the AP.

> With the advent of Wi-Fi 6, an alternative power save mechanism, known as Target Wake Time (TWT), has been introduced .

Author: Need to go through the whole doc and add quite a few terms. Not done yet.

This mechanism also enables STAs to enter a low power sleep state with receive traffic being buffered at the AP, but with much more flexibility in terms of the sleep interval.

Physical layer

> ==============

Reviewer: ```suggestion
--------------
```

> The Physical (PHY) layer is the final layer before the data is presented on the medium (radio frequency channel) and is primarily responsible for modulation and demodulation.

Reviewer: We've already introduced the term PHY in the baseband introduction.
```suggestion
The Physical layer is the final layer before the data is presented on the medium (radio frequency channel) and is primarily responsible for modulation and demodulation.
```
